src/grpc/mod.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2021 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::*;
+use futures::channel::oneshot;
+use futures::executor::block_on;
+use futures::prelude::*;
+use grpcio::{Environment, ServerBuilder};
+use std::io::Read;
+use std::net::IpAddr;
+use std::str;
+use std::sync::Arc;
+use std::{io, thread};
+
+pub mod protocol;
+
+use protocol::keyprovider;
+use protocol::keyprovider_grpc::*;
+use protocol::keyprovider_structs::*;
+
+use super::kbc_runtime;
+
+#[derive(Clone)]
+pub struct KeyProvider;
+
+impl KeyProviderService for KeyProvider {
+    fn un_wrap_key(
+        &mut self,
+        ctx: ::grpcio::RpcContext,
+        req: keyprovider::keyProviderKeyWrapProtocolInput,
+        sink: ::grpcio::UnarySink<keyprovider::keyProviderKeyWrapProtocolOutput>,
```

Comment:
It seems that grpcio generates traits for the grpc services with empty return values:

```Rust
pub trait KeyProviderService {
    fn wrap_key(&mut self, ctx: ::grpcio::RpcContext, req: super::keyprovider::keyProviderKeyWrapProtocolInput, sink: ::grpcio::UnarySink<super::keyprovider::keyProviderKeyWrapProtocolOutput>);
    fn un_wrap_key(&mut self, ctx: ::grpcio::RpcContext, req: super::keyprovider::keyProviderKeyWrapProtocolInput, sink: ::grpcio::UnarySink<super::keyprovider::keyProviderKeyWrapProtocolOutput>);                                                                                                                                                                                                                                                                                   Change Visibility to pub(crate)
}

```

Is there a way to make it generate traits that return a `Result`? And then propagate that back to the client?

Reply:
At present, grpcio crate does not support this, so after weighing and considering, I decided to use tonic crate to implement grpc service. Tonic has better robustness and performance than grpcio. In order to reduce AA footprint, ttrpc will be used instead of tonic grpc in the future.